u2f-ref-code/java/src/com/google/u2f/key/messages/TransferAccessMessage.java

```diff
@@ -0,0 +1,171 @@
+   * @param sequenceNumber: Int indicating position in the chain. 1 is first. 0 is reserved.
+   * @param newUserPublicKey: Public Key to which access is being transferred.
+   * @param applicationSha256
+   * @param newAttestationCertificate: Attestation cert of the key to which access is being
+   *        transferred
+   * @param signatureUsingAuthenticationKey: Signature using key from which access is being
+   *        transferred. Signature over sequenceNuber, access-transferred-from public key (old),
+   *        acces-transferred-to public key (newUserPublicKey), applicationSha256, and the
+   *        newAttestationCertificate.
+   * @param signatureUsingAttestationKey: Signature using the old (transferred access from)
+   *        attestation private key over all above parameters, including the
+   *        signatureUsingAuthenticationKey.
+   */
+  public TransferAccessMessage(int sequenceNumber, byte[] newUserPublicKey,
+      byte[] applicationSha256, X509Certificate newAttestationCertificate,
+      byte[] signatureUsingAuthenticationKey, byte[] signatureUsingAttestationKey) {
+    
+    Preconditions.checkNotNull(sequenceNumber, "Sequence Number should not be null");
+    Preconditions.checkNotNull(newUserPublicKey, "New user public key should not be null");
+    Preconditions.checkNotNull(applicationSha256, "Application Sha256 should not be null");
+    Preconditions.checkNotNull(newAttestationCertificate,
+        "New attestation certificate should not be null");
+    Preconditions.checkNotNull(signatureUsingAuthenticationKey,
+        "Signature using authentication key should not be null");
+    Preconditions.checkNotNull(signatureUsingAttestationKey,
+        "Signature using attestation Key should not be null");
+    
+    this.sequenceNumber = sequenceNumber;
+    this.newUserPublicKey = newUserPublicKey;
+    this.applicationSha256 = applicationSha256;
+    this.newAttestationCertificate = newAttestationCertificate;
+    this.signatureUsingAuthenticationKey = signatureUsingAuthenticationKey;
+    this.signatureUsingAttestationKey = signatureUsingAttestationKey;
+  }
+  
+  /**
+   * Build TransferAccessMessage from raw byte  array.
+   */
+  public static TransferAccessMessage fromBytes(byte[] data) throws U2FException {
+    try {
+      DataInputStream inputStream = new DataInputStream(new ByteArrayInputStream(data));
+
+      int sequenceNumber = inputStream.readUnsignedByte();
+      
+      byte[] newUserPublicKey = new byte[65];
+      inputStream.readFully(newUserPublicKey);
+      
+      byte[] applicationSha256 = new byte[32];
+      inputStream.readFully(applicationSha256);
+      
+      // TODO(alextaka): This doesn't seem to read the correct number of bytes
+      X509Certificate newAttestationCertificate = (X509Certificate) CertificateFactory.getInstance(
+          "X.509").generateCertificate(inputStream);
+      
+      byte[] signatureUsingAuthenticationKey = new byte[inputStream.readUnsignedByte()];
```

Comment:
give the signature length a symbolic name.  For example:

`byte signatureLength = inputStream.readUnsignedByte();`

that way the next line will be more readable.